src/main/java/edu/cornell/kfs/coa/service/impl/GlobalObjectWithIndirectCostRecoveryAccountsServiceImpl.java

```diff
@@ -0,0 +1,91 @@
+package edu.cornell.kfs.coa.service.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.kuali.kfs.coa.businessobject.IndirectCostRecoveryAccount;
+import org.kuali.rice.krad.bo.GlobalBusinessObjectDetailBase;
+
+import edu.cornell.kfs.coa.businessobject.GlobalObjectWithIndirectCostRecoveryAccounts;
+import edu.cornell.kfs.coa.businessobject.IndirectCostRecoveryAccountChange;
+import edu.cornell.kfs.coa.service.GlobalObjectWithIndirectCostRecoveryAccountsService;
+
+public class GlobalObjectWithIndirectCostRecoveryAccountsServiceImpl implements GlobalObjectWithIndirectCostRecoveryAccountsService{
+	
+    public List<IndirectCostRecoveryAccountChange> getActiveIndirectCostRecoveryAccounts(GlobalObjectWithIndirectCostRecoveryAccounts globalObjectWithIndirectCostRecoveryAccounts) {
+        List<IndirectCostRecoveryAccountChange> activeList = new ArrayList<IndirectCostRecoveryAccountChange>();
+        for (IndirectCostRecoveryAccountChange icr : globalObjectWithIndirectCostRecoveryAccounts.getIndirectCostRecoveryAccounts()){
+            if (icr.isActive()){
+                activeList.add(IndirectCostRecoveryAccountChange.copyICRAccount(icr));
+            }
+        }
+        return activeList;
+    }
+    
+    public void updateIcrAccounts(GlobalObjectWithIndirectCostRecoveryAccounts globalObjectWithIndirectCostRecoveryAccounts, GlobalBusinessObjectDetailBase globalDetail, List<IndirectCostRecoveryAccount> icrAccounts) {
+		Map<Integer, Integer> alreadyUpdatedIndexes = new HashMap<Integer, Integer>();
+		List<IndirectCostRecoveryAccount> addList = new ArrayList<IndirectCostRecoveryAccount>();
+		List<IndirectCostRecoveryAccountChange> indirectCostRecoveryAccounts = globalObjectWithIndirectCostRecoveryAccounts.getIndirectCostRecoveryAccounts();
+
+		if (indirectCostRecoveryAccounts.size() > 0) {
+			for (IndirectCostRecoveryAccountChange newICR : indirectCostRecoveryAccounts) {
+				boolean foundMatch = false;
+
+				int positionForMatchWithSameActiveIndicator = -1;
+				int currentPosition = 0;
+				int maxPosition = icrAccounts.size();
+
+				while (noMatchFoundOrSameActiveIndicatorMatchFoundAndStillHaveICRAccountToCheck(currentPosition, maxPosition, foundMatch, positionForMatchWithSameActiveIndicator)) {
+					if (!alreadyUpdatedIndexes.containsKey(currentPosition)) {
+						IndirectCostRecoveryAccount existingICR = icrAccounts.get(currentPosition);
+						// check if we have a match on chart, account and percentage
+						if (newICR.matchesICRAccount(existingICR)) {
+							// set this to true if we have found a match
```

Comment:
I'm not sure these comments, this one especially, are adding much value, and I would be inclined to remove them.